Failure in CREATE TABLE AS leaves inconsistent transaction if some rows were successfully inserted

Hello,
We have a rather annoying bug in version 1.4.200 (bp not reproduced in version 1.4.199).
Follow the steps below to reproduce it :

1. Create a txt file (C:\tmp\datafile.txt) with exactly this content :
```
column1;column2
;Lorem ipsum dolor
Lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor
```

2. Execute the SQL request to create a table from this file  :
`CREATE TABLE TABLE_FROM_FILE (column1 varchar(20), column2 varchar(20)) AS SELECT * FROM CSVREAD('C:\tmp\datafile.txt', 'column1;column2', 'charset=UTF-8 fieldDelimiter= lineSeparator=\n fieldSeparator=; writeColumnHeader=false' );
`
It will returned the error :
`Error: Value too long for column "column1 VARCHAR(20)": "'Lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor' (76)"; SQL statement:
CREATE TABLE TABLE_FROM_FILE (column1 varchar(20), column2 varchar(20)) AS SELECT * FROM CSVREAD('C:\tmp\datafile.txt', 'column1;column2', 'charset=UTF-8 fieldDelimiter= lineSeparator=\n fieldSeparator=; writeColumnHeader=false' ) [22001-200]
SQLState:  22001
ErrorCode: 22001`


3. Try to execute a 'CREATE TABLE' query  :
`CREATE TABLE TABLE_TEST (col1 varchar(20), col2 varchar(20));`

It will returned an inexpected error !
`Error: File corrupted while reading record: null. Possible solution: use the recovery tool; SQL statement:
CREATE TABLE TABLE_TEST (col1 varchar(20), col2 varchar(20)) [90030-200]
SQLState:  90030
ErrorCode: 90030`

So the problem is that we cannot execute any 'CREATE TABLE' and must restart the database engine to resolve this.

Best regards;

Thanks!

This issue can be reproduced with the following reduced test case in current H2:
```Java
try (Connection connection = DriverManager.getConnection("jdbc:h2:mem:");
        Statement statement = connection.createStatement()) {
    try {
        statement.execute("CREATE TABLE TEST(C VARCHAR(2)) AS VALUES 'A', 'AAA'");
    } catch (Exception e) {
    }
    statement.execute("CREATE TABLE TEST(C VARCHAR(2))");
}
```
```
Exception in thread "main" org.h2.jdbc.JdbcSQLNonTransientException:
General error: "An old transaction with the same id is still open: 1 [1.4.200/102]";
SQL statement:
CREATE TABLE TEST(C VARCHAR(2)) [50000-200]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:511)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:435)
	at org.h2.message.DbException.get(DbException.java:189)
	at org.h2.mvstore.db.Store.convertMVStoreException(Store.java:166)
	at org.h2.mvstore.db.MVTable.convertException(MVTable.java:679)
	at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:128)
	at org.h2.mvstore.db.MVTable.addRow(MVTable.java:506)
	at org.h2.engine.Database.addMeta(Database.java:905)
	at org.h2.engine.Database.addSchemaObject(Database.java:1102)
	at org.h2.command.ddl.CreateTable.update(CreateTable.java:123)
	at org.h2.command.CommandContainer.update(CommandContainer.java:173)
	at org.h2.command.Command.executeUpdate(Command.java:250)
	at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:232)
	at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:203)
        ....
Caused by: org.h2.mvstore.MVStoreException:
An old transaction with the same id is still open: 1 [1.4.200/102]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:1001)
	at org.h2.mvstore.tx.TransactionStore.addUndoLogRecord(TransactionStore.java:444)
	at org.h2.mvstore.tx.Transaction.log(Transaction.java:409)
	at org.h2.mvstore.tx.TxDecisionMaker.logAndDecideToPut(TxDecisionMaker.java:158)
	at org.h2.mvstore.tx.TxDecisionMaker$PutIfAbsentDecisionMaker.decide(TxDecisionMaker.java:288)
	at org.h2.mvstore.tx.TxDecisionMaker.decide(TxDecisionMaker.java:1)
	at org.h2.mvstore.MVMap$DecisionMaker.decide(MVMap.java:1683)
	at org.h2.mvstore.MVMap.operate(MVMap.java:1757)
	at org.h2.mvstore.tx.TransactionMap.set(TransactionMap.java:350)
	at org.h2.mvstore.tx.TransactionMap.putIfAbsent(TransactionMap.java:282)
	at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:115)
	... 9 more

```

--- a/h2/src/main/org/h2/mvstore/tx/TransactionStore.java
+++ b/h2/src/main/org/h2/mvstore/tx/TransactionStore.java
@@ -501,7 +501,7 @@ public class TransactionStore {
                     Record<?,?> op = cursor.getValue();
                     int mapId = op.mapId;
                     MVMap<Object, VersionedValue<Object>> map = openMap(mapId);
-                    if (map != null) { // might be null if map was removed later
+                    if (map != null && !map.isClosed()) { // might be null if map was removed later
                         Object key = op.key;
                         commitDecisionMaker.setUndoKey(undoKey);
                         // second parameter (value) is not really
@@ -509,9 +509,12 @@ public class TransactionStore {
                         map.operate(key, null, commitDecisionMaker);
                     }
                 }
-                undoLog.clear();
             } finally {
-                flipCommittingTransactionsBit(transactionId, false);
+                try {
+                    undoLog.clear();
+                } finally {
+                    flipCommittingTransactionsBit(transactionId, false);
+                }
             }
         }
     }

--- a/h2/src/test/org/h2/test/store/TestTransactionStore.java
+++ b/h2/src/test/org/h2/test/store/TestTransactionStore.java
@@ -24,6 +24,7 @@ import org.h2.mvstore.tx.TransactionMap;
 import org.h2.mvstore.tx.TransactionStore;
 import org.h2.mvstore.tx.TransactionStore.Change;
 import org.h2.mvstore.type.LongDataType;
+import org.h2.mvstore.type.StringDataType;
 import org.h2.store.fs.FileUtils;
 import org.h2.test.TestBase;
 import org.h2.util.Task;
@@ -935,4 +936,26 @@ public class TestTransactionStore extends TestBase {
         ts.close();
     }
 
+    private void testCommitAfterMapRemoval() {
+        try (MVStore s = MVStore.open(null)) {
+            TransactionStore ts = new TransactionStore(s);
+            ts.init();
+            Transaction t = ts.begin();
+            TransactionMap<Long,String> map = t.openMap("test", LongDataType.INSTANCE, StringDataType.INSTANCE);
+            map.put(1L, "A");
+            s.removeMap("test");
+            try {
+                t.commit();
+            } finally {
+                // commit should not fail, but even if it does
+                // transaction should be cleanly removed and store remains operational
+                assertTrue(ts.getOpenTransactions().isEmpty());
+                assertFalse(ts.hasMap("test"));
+                t = ts.begin();
+                map = t.openMap("test", LongDataType.INSTANCE, StringDataType.INSTANCE);
+                assertTrue(map.isEmpty());
+                map.put(2L, "B");
+            }
+        }
+    }
 }

